### PR TITLE
feat(new-integration): onboarding yandex.metrica destination

### PR DESCRIPTION
**Pull request by ujjwal-ab**
## Description of the change

> onboarding yandex.metrica destination

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

- - -
This change is [Reviewable](https://reviewable.io/reviews/rudderlabs/rudder-sdk-js/671)

### CODE REVIEW

File: integrations/YandexMetrica/browser.js [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        };
+      m[i].l = 1 * new Date();
+      for (var j = 0; j < document.scripts.length; j++) {
+        if (document.scripts[j].src === r) {
+          return;
+        }
+      }
+      (k = e.createElement(t)),
+        (a = e.getElementsByTagName(t)[0]),
+        (k.async = 1),
```
Comment (MoumitaM) [inline integrations/YandexMetrica/browser.js:42]: Add this:
import { LOAD_ORIGIN } from '../ScriptLoader';
k.setAttribute('data-loader', LOAD_ORIGIN),

File: integrations/YandexMetrica/browser.js [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+  eventMapping,
+} from "./utils";
+import {
+  removeUndefinedAndNullValues,
+  getHashFromArrayWithDuplicate,
+} from "../utils/commonUtils";
+import { NAME } from "./constants";
+
+class YandexMetrica {
+  constructor(config) {
```
Comment (MoumitaM) [inline integrations/YandexMetrica/browser.js:16]:
```suggestion
  constructor(config, analytics) {
```

File: integrations/YandexMetrica/browser.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    this.tagId = config.tagId;
+    this.clickMap = config.clickMap;
+    this.trackLinks = config.trackLinks;
+    this.trackBounce = config.trackBounce;
+    this.webvisor = config.webvisor;
+    this.containerName = config.containerName;
+    this.eventNameToYandexEvent = config.eventNameToYandexEvent;
+    this.name = NAME;
+  }
```
Comment (MoumitaM) [inline integrations/YandexMetrica/browser.js:25]: Add this inside constructor:
```
if (analytics.logLevel) {
      logger.setLogLevel(analytics.logLevel);
    }
```

File: integrations/YandexMetrica/browser.js [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+    return !!window.ym && typeof window.ym === "function";
+  }
+
+  isReady() {
+    logger.debug("===In isReady YandexMetrica===");
+    return !!window.ym;
+  }
+
+  identify(rudderElement) {
+    logger.debug("===In YandexMetrica Identify");
```
Comment (ItsSudip) [inline integrations/YandexMetrica/browser.js:74]:
```suggestion
    logger.debug("===In YandexMetrica Identify===");
```

File: integrations/YandexMetrica/utils.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      if (!(properties.product_id || properties.name)) {
+        logger.error(`None of product_id or name is present for the product`);
+      } else {
+        responsePayload = populatePayload(
+          eventMapping[trimmedEvent],
+          properties
+        );
+      }
+      break;
+    }
```
Comment (ItsSudip) [inline integrations/YandexMetrica/utils.js:125]: Can't we merge these three cases?

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: integrations/YandexMetrica/utils.js [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+      revenue: parseFloat(properties.revenue),
+    };
+    payload.ecommerce[eventType].actionField = actionField;
+  }
+  return payload;
+};
+
+const commonEventPayload = (eventType, properties) => {
+  let responsePayload = {};
+  if (!(properties.product_id || properties.name)) {
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply integrations/YandexMetrica/utils.js:58]: for the default e-comm events, we are checking inside the products as we have that kind of structure for those events and this is for any other custom event apart from those. Is this check required for other events too?

File: integrations/YandexMetrica/browser.js [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+    const { properties } = message;
+    if (!properties.url) {
+      logger.error("[Yandex Metrica]: url from page call is missing!!===");
+      return;
+    }
+    let payload = {};
+    payload = {
+      ...payload,
+      title: properties.title,
+      referer: properties.referrer,
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js:143]: That is not a required field but in that case, the Page title and URL are automatically mapped into that.

File: integrations/YandexMetrica/browser.js [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+        (a = e.getElementsByTagName(t)[0]),
+        (k.async = 1),
+        (k.src = r),
+        a.parentNode.insertBefore(k, a);
+    })(window, document, "script", "https://mc.yandex.ru/metrika/tag.js", "ym");
+
+    ym(this.tagId, "init", {
+      clickmap: this.clickMap,
+      trackLinks: this.trackLinks,
+      accurateTrackBounce: this.accurateTrackBounce,
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js:50]: Yes, taking it. There is a toggle for this there.

File: integrations/YandexMetrica/browser.js (hunk 1 above)
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js:42]: done

File: integrations/YandexMetrica/browser.js (hunk 2 above)
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js:16]: Done

File: integrations/YandexMetrica/browser.js (hunk 3 above)
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js:25]: Added

File: integrations/YandexMetrica/browser.js [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+          return;
+        }
+      }
+      (k = e.createElement(t)),
+        (a = e.getElementsByTagName(t)[0]),
+        (k.async = 1),
+        (k.src = r),
+        a.parentNode.insertBefore(k, a);
+    })(window, document, "script", "https://mc.yandex.ru/metrika/tag.js", "ym");
+
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js:46]: Added

File: integrations/YandexMetrica/browser.js [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+    ym(this.tagId, "init", {
+      clickmap: this.clickMap,
+      trackLinks: this.trackLinks,
+      accurateTrackBounce: this.accurateTrackBounce,
+      webvisor: this.webvisor,
+      ecommerce: this.containerName,
```
[the remark this answers is not in the document]
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js:52]: Done

File: integrations/YandexMetrica/utils.js (hunk 5 above)
Comment (ujjwal-ab) [reply integrations/YandexMetrica/utils.js:125]: Merged

File: integrations/YandexMetrica/utils.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (ItsSudip) [reply integrations/YandexMetrica/utils.js:58]: Yes this check is required.

File: integrations/YandexMetrica/utils.js [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+      coupon: properties.coupon,
+      goal_id: parseInt(properties.goal_id, 10),
+      revenue: parseFloat(properties.revenue),
+    };
+    payload.ecommerce[eventType].actionField = actionField;
+  }
+  return payload;
+};
+
+const commonEventPayload = (eventType, properties) => {
```
Comment (ItsSudip) [inline integrations/YandexMetrica/utils.js:56]: `eventType` is an array here as we are using `getHashFromArrayWithDuplicate`.

Comment (ujjwal-ab) [reply integrations/YandexMetrica/utils.js:56]: No, it is not an array and we are not using `getHashFromArrayWithDuplicate` for this.

File: integrations/YandexMetrica/browser.js [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+    this.tagId = config.tagId;
+    this.clickMap = config.clickMap;
+    this.trackLinks = config.trackLinks;
+    this.trackBounce = config.trackBounce;
+    this.webvisor = config.webvisor;
+    this.containerName = config.containerName;
+    this.goalId = config.goalId;
+    if (!config.containerName) {
+      this.containerName = "dataLayer";
+    }
```
Comment (ItsSudip) [inline integrations/YandexMetrica/browser.js:24]:
```suggestion
    this.containerName = config.containerName || "dataLayer";
    this.goalId = config.goalId;
```

**Review by ItsSudip — COMMENTED** [review]
Add comments before the function definition.

File: integrations/YandexMetrica/browser.js (hunk 12 above)
Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js]: Done

**Comment by sauravlal-rudderstack** [comment]
LGTM

File: integrations/YandexMetrica/browser.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+import {
+  removeUndefinedAndNullValues,
+  getHashFromArrayWithDuplicate,
+} from "../utils/commonUtils";
+import { NAME } from "./constants";
+import { LOAD_ORIGIN } from "../ScriptLoader";
+
+class YandexMetrica {
+  constructor(config, analytics) {
+    this.analytics = analytics;
```
Comment (MoumitaM) [inline integrations/YandexMetrica/browser.js:13]: Line 13 is not needed

Comment (ujjwal-ab) [reply integrations/YandexMetrica/browser.js]: removed

**Review by MoumitaM — APPROVED** [review]

**Review by MoumitaM — DISMISSED** [review]

**Review by utsabc — APPROVED** [review]

**Review by MoumitaM — APPROVED** [review]

**Review by ItsSudip — APPROVED** [review]

**Pull request merged**